File: frontend/shader/contract/ts/cardano.ts
// ============================================================================
// Cardano Configuration
// ============================================================================

export interface CardanoConfig {
  rpcUrl: string;
  network: "mainnet" | "testnet" | "preview" | "preprod";
}

// ============================================================================
// Cardano API Response Types
// ============================================================================

export interface CardanoAddressResponse {
  amount?: Array<{
    unit: string;
    quantity: string;
  }>;
}

/**
 * Cardano Connector
 */
export class CardanoConnector {
  private readonly config: CardanoConfig;

  constructor(config: CardanoConfig) {
    this.config = config;
  }

  async getBalance(address: string): Promise<string> {
    if (!this.isValidAddress(address)) {
      throw new Error("Invalid Cardano address");
    }

    try {
      let apiUrl: string;
      if (this.config.network === "mainnet") {
        apiUrl = `https://cardano-mainnet.blockfrost.io/api/v0/addresses/${address}`;
      } else if (this.config.network === "testnet") {
        apiUrl = `https://cardano-testnet.blockfrost.io/api/v0/addresses/${address}`;
      } else {
        apiUrl = `${this.config.rpcUrl}/addresses/${address}`;
      }

      const response = await fetch(apiUrl, {
        headers: {
          "Content-Type": "application/json",
          // Note: In production, add project_id header for Blockfrost API
        },
      });

      if (!response.ok) {
        throw new Error("Failed to fetch Cardano balance");
      }

      const data = (await response.json()) as CardanoAddressResponse;
      // Return balance in lovelace (smallest unit)
      return (
        data.amount?.find((item: any) => item.unit === "lovelace")?.quantity ||
        "0"
      );
    } catch (error) {
      throw new Error(
        `Cardano getBalance failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  async sendTransaction(
    fromAddress: string,
    toAddress: string,
    amount: string
  ): Promise<string> {
    if (!this.isValidAddress(fromAddress) || !this.isValidAddress(toAddress)) {
      throw new Error("Invalid Cardano address");
    }

    const amountNum = Number.parseInt(amount, 10);
    if (Number.isNaN(amountNum) || amountNum <= 0) {
      throw new Error("Invalid amount");
    }

    try {
      // Cardano transactions require:
      // 1. Building the transaction with proper UTXOs
      // 2. Signing with the private key
      // 3. Submitting to the network
      // This is a simplified version - in production use cardano-serialization-lib or similar

      let apiUrl: string;
      if (this.config.network === "mainnet") {
        apiUrl = "https://cardano-mainnet.blockfrost.io/api/v0";
      } else if (this.config.network === "testnet") {
        apiUrl = "https://cardano-testnet.blockfrost.io/api/v0";
      } else {
        apiUrl = this.config.rpcUrl;
      }

      // Get UTXOs for fromAddress
      const utxoResponse = await fetch(
        `${apiUrl}/addresses/${fromAddress}/utxos`,
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );

      if (!utxoResponse.ok) {
        throw new Error("Failed to fetch UTXOs");
      }

      // Build and submit transaction
      // Note: In production, use proper transaction building library
      const submitResponse = await fetch(`${apiUrl}/tx/submit`, {
        method: "POST",
        headers: {
          "Content-Type": "application/cbor",
        },
        body: JSON.stringify({
          // Transaction CBOR would go here
          // This requires proper serialization
        }),
      });

      if (!submitResponse.ok) {
        throw new Error("Failed to submit Cardano transaction");
      }

      const submitData = (await submitResponse.json()) as { hash?: string };
      return submitData.hash || "transaction_hash";
    } catch (error) {
      throw new Error(
        `Cardano sendTransaction failed: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  isValidAddress(address: string): boolean {
    return this.config.network === "mainnet"
      ? address.startsWith("addr1")
      : address.startsWith("addr_test");
  }

  /**
   * Bridge tokens from Cardano to another chain
   */
  async bridgeTokens(
    sourceAddress: string,
    destinationAddress: string,
    amount: string
  ): Promise<{ sourceTx: string; destinationTx?: string }> {
    const txHash = await this.sendTransaction(
      sourceAddress,
      destinationAddress,
      amount
    );
    return { sourceTx: txHash };
  }

  /**
   * Get chain name for bridge operations
   */
  getChainName(): string {
    return "Cardano";
  }

  /**
   * Get chain aliases for bridge operations
   */
  getChainAliases(): string[] {
    return ["cardano", "ada"];
  }
}

File: store/rust/rust/cardano.rs
//! Cardano connector for interacting with Cardano blockchain

use crate::{http_get_json, http_post_json, BridgeError, BridgeResult, CardanoNetwork};
use serde::{Deserialize, Serialize};

/// Cardano connector for interacting with Cardano blockchain
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CardanoConnector {
    rpc_url: String,
    network: CardanoNetwork,
}

impl CardanoConnector {
    pub fn new(rpc_url: &str, network: CardanoNetwork) -> BridgeResult<Self> {
        Ok(Self { rpc_url: rpc_url.to_string(), network })
    }

    pub fn get_balance(&self, address: &str) -> BridgeResult<u64> {
        if !self.is_valid_address(address) {
            return Err(BridgeError::InvalidAddress(address.to_string()));
        }

        // Cardano uses Ogmios or Blockfrost API
        // Try Blockfrost-style API first
        let url = format!("{}/addresses/{}", self.rpc_url.trim_end_matches('/'), address);
        match http_get_json(&url) {
            Ok(result) => {
                // Blockfrost format: {"amount": [{"unit": "lovelace", "quantity": "..."}]}
                let amount = result
                    .get("amount")
                    .and_then(|a| a.as_array())
                    .and_then(|arr| {
                        arr.iter()
                            .find(|item| item.get("unit").and_then(|u| u.as_str()) == Some("lovelace"))
                            .and_then(|item| item.get("quantity"))
                            .and_then(|q| q.as_str())
                            .and_then(|s| s.parse::<u64>().ok())
                    })
                    .unwrap_or(0);
                Ok(amount)
            },
            Err(_) => {
                // Fallback: try Ogmios-style query
                let params = serde_json::json!({"address": address});
                let result = http_post_json(&self.rpc_url, params)?;
                result
                    .get("lovelace")
                    .and_then(|l| l.as_u64())
                    .ok_or_else(|| BridgeError::NetworkError("Failed to get Cardano balance".to_string()))
            },
        }
    }

    pub fn send_transaction(
        &self,
        from_address: &str,
        to_address: &str,
        amount: u64, // in lovelace
    ) -> BridgeResult<String> {
        if !self.is_valid_address(from_address) || !self.is_valid_address(to_address) {
            return Err(BridgeError::InvalidAddress("Invalid Cardano address".to_string()));
        }
        if amount == 0 {
            return Err(BridgeError::InvalidAmount("Amount must be greater than 0".to_string()));
        }

        // Cardano transactions require building, signing, and submitting
        // This is a simplified version that assumes the RPC endpoint handles transaction building
        let body = serde_json::json!({
            "from": from_address,
            "to": to_address,
            "amount": amount.to_string()
        });

        let result = http_post_json(&self.rpc_url, body)?;
        result
            .get("txHash")
            .or_else(|| result.get("tx_hash"))
            .and_then(|h| h.as_str())
            .map(|s| s.to_string())
            .ok_or_else(|| BridgeError::TransactionFailed("Failed to get transaction hash".to_string()))
    }

    pub fn is_valid_address(&self, address: &str) -> bool {
        // Cardano addresses start with 'addr' for mainnet or 'addr_test' for testnet
        match self.network {
            CardanoNetwork::Mainnet => address.starts_with("addr1"),
            CardanoNetwork::Testnet | CardanoNetwork::Preview | CardanoNetwork::Preprod => {
                address.starts_with("addr_test")
            },
        }
    }
}
